src/rgw/rgw_op.cc

```diff
@@ -4003,6 +4020,13 @@ void RGWPutObj::execute(optional_yield y)
     std::unique_ptr<rgw::sal::MultipartUpload> upload;
     upload = s->bucket->get_multipart_upload(s->object->get_name(),
 					 multipart_upload_id);
+    rgw::sal::Attrs mpattrs;
+    op_ret = upload->get_info(this, s->yield, nullptr, &mpattrs);
```

Comment:
i'm concerned about this additional rados read for the multipart upload info; this is a cost we'll have to pay for all part uploads, whether they're checksummed or not

assuming that each part upload is providing its own `x-amz-checksum-` header, we should checksum based on that instead of the attr from `CreateMultipartUpload`. then on `CompleteMultipartUpload` when calculating the checksum-of-checksums (not implemented here?), we can verify that all of its parts use the same algorithm as `CreateMultipartUpload`

we use a similar strategy for compression types in https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_sal_rados.cc#L2612-L2617

Reply:
Why it is implemented here , is , (as AWS does) each part is to be checked whether x-amz-checksum- header is carried if an algorithm is set when creating MPupload (otherwise for this part response 400 EINVAL).

Reply:
thanks @imtzw. i think it's probably worth breaking from AWS' behavior in this case. there's a clear performance cost, and i don't see much benefit

if the client sends a different algorithm for PutObject vs CreateMultipartUpload, isn't that just a bug in the client? does it really make a difference whether we reject the PutObject instead of the final CompleteMultipartUpload?

Reply:
i think i was mistaken on this part. i now see that we're already calling this just below:
> op_ret = upload->get_info(this, s->yield, &pdest_placement);